lib/util.h

```diff
@@ -129,6 +129,20 @@ size_t idx_1st_upper_bound(const double *values, size_t n_values, double query);
 size_t idx_1st_strict_upper_bound(
     const double *elements, size_t n_elements, double query);
 
+inline const double *
+ptr_1st_upper_bound(const double *start, const double *stop, double query)
```

Comment:
I'm remembering now that in another PR you had a similar reaction to inline functions in headers. For this PR I'll just avoid inline functions in headers like I did in that other PR.

I'm guessing that, at least partly, one of the things you don't like is having function implementations mixed in with function declarations; breaking up what would otherwise be a single coherent block of just declarations.

I can make a branch, kind of like what I did with the macro, with a suggestion for keeping a "clean" section of just declarations, and then inline implementations in a separate section at the very end of the header file (or it can be a separate 'inline implementations' file). Or perhaps you already know about this approach and already know you do not like it. I can include some thoughts on the suggestion and also on how the inline keyword is currently being used in the codebase.

Regarding the difference it makes in terms of compiled code I have a complicated answer, but the simple answer is that it is not measurable in time with my current test setup (which has high variance in timings).

Reply:
The main thing I'm aiming for is simplicity: someone who knows a little bit of C should be able to read this and understand it *unless* there's a good reason for doing tricky complicated things. If we can't measure the difference perf-wise, then that's not a good reason in my book.